File: week_2/day_08_stack/variations_que/minStack.js
// minStack.js
class MinStack {
    constructor() {
        this.stack = [];
    }
    push(val) {
        const minVal = this.stack.length === 0 ? val : Math.min(val, this.stack[this.stack.length - 1][1]);
        this.stack.push([val, minVal]);
    }
    pop() { this.stack.pop(); }
    top() { return this.stack[this.stack.length - 1][0]; }
    getMin() { return this.stack[this.stack.length - 1][1]; }
}

// INPUT
const ms = new MinStack();
ms.push(3); ms.push(5); ms.push(2);
console.log(ms.getMin()); // 2

File: week_2/day_08_stack/variations_que/min_stack.py
# min_stack.py
class MinStack:
    def __init__(self):
        self.stack = []

    def push(self, val):
        min_val = val if not self.stack else min(val, self.stack[-1][1])
        self.stack.append((val, min_val))

    def pop(self):
        self.stack.pop()

    def top(self):
        return self.stack[-1][0]

    def getMin(self):
        return self.stack[-1][1]

# INPUT
ms = MinStack()
ms.push(3); ms.push(5); ms.push(2)
print(ms.getMin())  # 2
